core/src/main/java/org/apache/calcite/util/SealableMap.java

```diff
@@ -0,0 +1,180 @@
+
+import com.google.common.collect.ForwardingMap;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+
+/**
+ * This implementation of {@link Map} can be sealed and unsealed at will. When
+ * sealed, it does not accept any writes.
+ *
+ * @param <K> the type of keys maintained by this map
+ * @param <V> the type of mapped values
+ */
+public class SealableMap<K, V> extends ForwardingMap<K, V> {
+  //~ Instance fields --------------------------------------------------------
+
+  private final Map<K, V> delegate;
+  private final Map<K, V> readOnlyDelegate;
+  private boolean sealed = false;
+
+  //~ Constructors -----------------------------------------------------------
+
+  /**
+   * Creates a map that can be sealed and unsealed at will. When sealed, the
+   * map will not accept any write operations
+   * @param delegate the map that all calls will be delegated to. Will not
+   *                 delegate write calls when sealed. Any changes
+   *                 to the delegate will reflect in the {@link SealableMap}
+   */
+  public SealableMap(Map<K, V> delegate) {
+    this.delegate = Objects.requireNonNull(delegate, "Delegate map cannot be null");
+    this.readOnlyDelegate = Collections.unmodifiableMap(delegate);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  // SealableMap public methods
+
+  /**
+   * Seals the map for writes. No effect if already sealed
+   */
+  public void seal() {
+    sealed = true;
+  }
+
+  /**
+   * Unseals the map for writes. No effect if already unsealed
+   */
+  public void unseal() {
+    sealed = false;
+  }
+
+  /**
+   * @return if the map is currently sealed
+   */
+  public boolean isSealed() {
+    return sealed;
+  }
+
+  // Implemented for ForwardingMap
+
+  @Override
+  protected Map<K, V> delegate() {
+    return sealed ? readOnlyDelegate : delegate;
+  }
+
+  // overridden from ForwardingMap
+
+  @Override public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof SealableMap)) {
+      return false;
+    }
+    SealableMap<?, ?> that = (SealableMap<?, ?>) o;
+    return sealed == that.sealed && super.equals(that.delegate);
+  }
+
+  @Override public int hashCode() {
+    return Objects.hash(delegate, sealed);
+  }
+
+  @Override public String toString() {
+    return "SealableMap{" + "delegate=" + delegate + ", sealed=" + sealed + '}';
+  }
+
+  // Overridden from Map
+
+  @Override public V getOrDefault(Object key, V defaultValue) {
+    return delegate.getOrDefault(key, defaultValue);
+  }
+
+  @Override public void forEach(BiConsumer<? super K, ? super V> action) {
+    delegate.forEach(action);
+  }
+
+  @Override public void replaceAll(BiFunction<? super K, ? super V, ? extends V> function) {
+    checkSealed();
+    delegate.replaceAll(function);
```

Comment:
why not do `super.replaceAll(function)`? similar comment on all the other methods related to updates in the map